### PR TITLE
Test macOS arm64 Python wheel using Fly CI

### LINKED ISSUE #476: [FEATURE] Enable native tests on M1 Mac using FlyCI
# Introduction

GitHub Actions does not have free M1 Mac instance. We now build our Mac M1 in `x64` but cannot test them.

# Proposal

[FlyCI](https://www.flyci.net/) provides some free minutes for M1 Mac. It is limited. But we can use them. Maybe we can enable native build and test of `arm64` Mac on `main` for Python packages.

# Usage of runs per month

Bellow is a summary of runs in `main.yml` per month. We had100+ successful runs sometimes per month. Therefore, we cannot afford to always run FlyCI. We can only run on `main`.

```
                     fail  success  cancelled
timestamp timestamp                          
2022      12            4       58          0
2023      1            54      148          1
          2            30      100          1
          3            21       91          2
          4            62      115          1
          5            13       34          5
          6            52      174         32
          7           109      187         25
          8            50      134        110
          9            51       73         97
          10           33       69         89
          11           22       71         63
          12           22       43         52
2024      1            52       54         40
```

### PR DESCRIPTION
**Pull request by TonyXiang8787**
Closes #476 

# Approach

- Only test if we want to publish wheel, before the publish step. 
- We download the built `arm64` wheel, install it with dependencies and run tests.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -248,8 +248,43 @@ jobs:
       - name: Test
         run: pytest
 
+  test-macos-arm64:
+    runs-on: flyci-macos-large-latest-m1
```
Comment (TonyXiang8787) [inline .github/workflows/main.yml:252]: If things run correctly, I will add a condition here like
```
if: (github.event_name == 'push') || (github.event_name == 'workflow_dispatch')
```

Comment (TonyXiang8787) [reply .github/workflows/main.yml:252]: So it will run only if we publish

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -248,8 +248,43 @@ jobs:
       - name: Test
         run: pytest
 
+  test-macos-arm64:
```
Comment (mgovers) [inline .github/workflows/main.yml:251]: optionally: rename to `test-python-macos-arm64`

**Review by mgovers — DISMISSED** [review]
LGTM

**Comment by mgovers** [comment]
it might be worth doing one final `manual_dispatch` on this branch to verify that everything indeed works as intended

**Comment by TonyXiang8787** [comment]
> it might be worth doing one final `manual_dispatch` on this branch to verify that everything indeed works as intended

You can't. `manual_dispatch` uses the workflow file from `main` branch.

File: .github/workflows/main.yml (hunk 1 above)
Comment (mgovers) [reply .github/workflows/main.yml:252]: right... i forgot

**Review by mgovers — APPROVED** [review]

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -248,8 +248,47 @@ jobs:
...
+          pip install power-grid-model[dev]==$(cat PYPI_VERSION) --find-links=wheelhouse-macos
+      
+      - name: test
+        run: |
+          pytest
+
+
   publish-wheels:
-    needs: [build-cpp-test-linux, build-cpp-test-windows, build-cpp-test-macos, build-and-test-python, build-and-test-conda]
+    needs: [build-cpp-test-linux, build-cpp-test-windows, build-cpp-test-macos, build-and-test-python, build-and-test-conda, test-python-macos-arm64]
```
Comment (mgovers) [inline .github/workflows/main.yml:291]: publish wheels was skipped

**Review by mgovers — CHANGES_REQUESTED** [review]
publish wheels was skipped in the last CI run. that's not good

File: .github/workflows/main.yml (hunk 3 above)
Comment (TonyXiang8787) [reply .github/workflows/main.yml:291]: Use the solution here:

https://stackoverflow.com/questions/69354003/github-action-job-fire-when-previous-job-skipped

**Comment by TonyXiang8787** [comment]
@mgovers with the new adjustment we can enable auto-merge again. The publish job will always run. But it will fail at the first step if previous jobs fail. This will produce a `failure` for the `publish` status check, instead of `skipped`. And the PR will be blocked.

**Comment by mgovers** [comment]
> @mgovers with the new adjustment we can enable auto-merge again. The publish job will always run. But it will fail at the first step if previous jobs fail. This will produce a `failure` for the `publish` status check, instead of `skipped`. And the PR will be blocked.

That's great news!

**Review by mgovers — APPROVED** [review]
Great!

**Comment by kgantchev** [comment]
Hi @TonyXiang8787,

Thanks for trusting [FlyCI](https://flyci.net) to handle your MacOS runs. Please let us know if you run into any issues. Our team is always here to help.

Best Regards,
Kiril Gantchev

Web: [flyci.net](https://flyci.net/) | Twitter: [flyciapp](https://twitter.com/flyciapp) | LinkedIn: [FlyCI](https://linkedin.com/company/flyci) | Discord: [FlyCI](https://discord.gg/JyCjh439da)

**Comment by kgantchev** [comment]
Please leave us a comment on what's your experience with FlyCI so far. **We'd love to get your feedback**!

**Comment by TonyXiang8787** [comment]
Hi @kgantchev,

The integration of FlyCI macOS runner in our workflow is seemless. We just had to change the `runs-on` attribute. The only thing we needed to think about is how and when we use the runner from FlyCI to be cost efficiently. So currently the job is only run in `main`.

Regards,
Tony Xiang

**Comment by kgantchev** [comment]
Hi @TonyXiang8787,

Thanks for continuing to use FlyCI. We would love to feature your feedback on our website and add PowerGridModel to [our case studies section](https://www.flyci.net/case-studies?utm_source=github&utm_medium=community&utm_campaign=recommend). It will highlight your project to our visitors and it will help us grow our community. 

I have a few quick questions that will help me prepare the case study: 

- **What motivated your choice to use FlyCI's MacOS runners, and what key benefits have you observed?** (Performance, compatibility, etc.)  
- **What has been your experience with the FlyCI runners so far?** 
- **What improvements have you noticed since integrating FlyCI's MacOS runners?** (e.g., performance enhancements, reduced build times, developer productivity, cost savings)

**Comment by TonyXiang8787** [comment]
Hi @kgantchev,

Thanks for your questions. See my answers below.

* Our main motivation is to test macOS arm 64 builds. Previously that was not possible and now we can test the build with FlyCI macOS runner.
* It runs as expected with seamless integration with GitHub Actions.
* The main improvement is that we can now test macOS arm64 build so we can make sure the package runs correctly in all release platforms.

**Comment by kgantchev** [comment]
Thanks Tony! Do you mind if I share this feedback on our Case Studies section? 

If you're OK with it, then I'll send you the draft for review.

**Comment by TonyXiang8787** [comment]
> Thanks Tony! Do you mind if I share this feedback on our Case Studies section?
> 
> If you're OK with it, then I'll send you the draft for review.

No problem.

**Comment by kgantchev** [comment]
Excellent! Thank you! I'll send you an update with the draft when we're ready.

**Pull request merged**